File: chapter03/python/pathSum.py
class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


# pathSumI会找到不连续的路径, 例如:10 -> -2, 所以行不通, 要保证, 一旦pathSumI(root.left, target - root.val), 则root.left后续每个节点都要减去val
def path_sum_bad(root, target):
    if root is None:
        return 0

    if target == root.val:
        return 1

    res = 0
    res += path_sum_bad(root.left, target)

    res += path_sum_bad(root.left, target - root.val)

    res += path_sum_bad(root.right, target)

    res += path_sum_bad(root.right, target - root.val)

    return res


def count(root, target):
    if root is None:
        return 0
    return (root.val == target) + count(root.left, target-root.val) + count(root.right, target-root.val)


def path_sum_good(root, target):
    if root is None:
        return 0
    return count(root, target) + count(root.left, target) + count(root.right, target)


if __name__ == "__main__":
    target = 8

    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(-3)

    root.left.left = TreeNode(3)
    root.left.right = TreeNode(2)
    root.right.right = TreeNode(11)

    root.left.left.left = TreeNode(-1)
    root.left.left.right = TreeNode(-2)
    root.left.right.right = TreeNode(1)

    print(path_sum_good(root, target))

File: chapter03/cpp/pathSum.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int _val=0, TreeNode* _left=nullptr, TreeNode* _right=nullptr) : val(_val), left(_left), right(_right) {}
};

int path_sum_helper(TreeNode* root, int target)
{
    if(root == nullptr)
        return 0;

    return (target == root->val) + path_sum_helper(root->left, target-root->val) + path_sum_helper(root->right, target-root->val);
}

int path_sum(TreeNode* root, int target)
{   
    if(root == nullptr)
        return 0;
    
    return path_sum_helper(root, target) + path_sum(root->left, target) + path_sum(root->right, target);
    
}


int main()
{
    TreeNode* root = new TreeNode(10);
    root->left = new TreeNode(5);
    root->right = new TreeNode(-3);

    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(2);

    root->right->right = new TreeNode(11);

    root->left->left->left = new TreeNode(3);
    root->left->left->right = new TreeNode(-2);

    root->left->right->right = new TreeNode(1);

    int target = 8;
    cout << path_sum(root, target) << endl;
    return 0;
}